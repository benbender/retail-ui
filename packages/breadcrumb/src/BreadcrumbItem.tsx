import { Theme, useThemeCtx } from '@retail-ui/theme'
import { cloneElement } from '@retail-ui/utils'
import clsx from 'clsx'
import * as React from 'react'

type ReactLIProps = React.HTMLAttributes<HTMLLIElement>
type Ref = HTMLLIElement

type BreadcrumbItemSize = keyof Theme['BreadcrumbStyles']['item']['size']

export const BREADCRUMB_SEPARATOR_VARIANT = {
  default: '/',
  arrow: 'â†’',
  arrowhead: '>',
  bullet: '.',
}
export type BreadcrumbItemProps = {
  separator?: keyof typeof BREADCRUMB_SEPARATOR_VARIANT
  onClick?: () => void
  isActive?: boolean
  size?: BreadcrumbItemSize
  icon?: React.ReactNode
}

// TODO: add hover style for primary text
export const BreadcrumbItem = React.forwardRef<
  Ref,
  ReactLIProps & BreadcrumbItemProps
>((props, ref) => {
  const {
    children,
    separator = 'default',
    className,
    isActive = false,
    size = 'base',
    icon,
    ...rest
  } = props

  const {
    theme: { BreadcrumbStyles },
  } = useThemeCtx()

  const itemCls = BreadcrumbStyles.item

  const sizeCls = itemCls.size

  const cls = clsx(
    className,
    itemCls.base,
    isActive && itemCls.active,
    sizeCls[size],
  )

  let linkToRender
  if (icon) {
    const iconSizeCls = itemCls.icon.size
    const iconCls = clsx(itemCls.icon.base, iconSizeCls[size])
    const iconNode = cloneElement(icon, { className: iconCls })

    linkToRender = (
      <span className="flex items-center">
        {iconNode}
        {children}
      </span>
    )
  } else {
    linkToRender = <span>{children}</span>
  }

  const separatorCls = clsx(BreadcrumbStyles.separator.base)

  return (
    <React.Fragment>
      <li ref={ref} className={cls} {...rest}>
        {linkToRender}
      </li>
      {separator && (
        <span className={separatorCls}>
          {BREADCRUMB_SEPARATOR_VARIANT[separator]}
        </span>
      )}
    </React.Fragment>
  )
})

BreadcrumbItem.displayName = 'BreadcrumbItem'
