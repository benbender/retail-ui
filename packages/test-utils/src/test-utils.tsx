import { defaultTheme, ThemeProvider } from '@retail-ui/theme'
import {
  render as RtlRender,
  RenderOptions,
  RenderResult,
} from '@testing-library/react'
import * as React from 'react'

type UI = Parameters<typeof RtlRender>[0]

// UI-less passthrough fallback to prevent using conditional logic in render
function ChildrenPassthrough({ children }: { children: React.ReactElement }) {
  return children
}

export interface TestOptions extends Omit<RenderOptions, 'wrapper'> {
  /**
   * optional additional wrapper, e.g. Context
   *
   * @example
   * ```ts
   * // single wrapper
   * render(<MyConponent />, {
   *  wrapper: MyContext
   * });
   *
   * // multiple wrapper
   * render(<MyConponent />, {
   *  wrapper: ({ children }) => (
   *    <ContextA>
   *      <ContextB>
   *        {children}
   *      <ContextB />
   *    <ContextA />
   *  )
   * });
   *
   * ```
   */
  wrapper?: typeof ChildrenPassthrough
}

/**
 * Custom render for @testing-library/react
 *
 * @see https://testing-library.com/docs/react-testing-library/setup#custom-render
 * @param component the component under test
 * @param options customized test options
 */
export const render = (
  ui: UI,
  { wrapper: Wrapper = ChildrenPassthrough, ...options }: TestOptions = {},
): RenderResult => {
  return RtlRender(
    <ThemeProvider value={{ theme: defaultTheme }}>
      <Wrapper>{ui}</Wrapper>
    </ThemeProvider>,
    options,
  )
}

export { fireEvent } from '@testing-library/react'
