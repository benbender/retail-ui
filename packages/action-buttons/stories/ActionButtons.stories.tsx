import * as React from 'react'

import { ActionButtons, ActionButtonsProps } from '../src'

const meta = {
  title: 'Retail-UI/ActionButtons',
  component: ActionButtons,
}

export default meta

const Template = (args: ActionButtonsProps) => (
  <ActionButtons {...args} data-testid={meta.title} />
)

export const Simple = Template.bind({})
// @ts-ignore
Simple.args = {}
