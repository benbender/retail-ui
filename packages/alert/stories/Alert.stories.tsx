import * as React from 'react'

import { Alert, AlertProps } from '../src'

const meta = {
  title: 'Retail-UI/Alert',
  component: Alert,
}

export default meta

const Template = (args: AlertProps) => (
  <Alert {...args} data-testid={meta.title} />
)

export const Simple = Template.bind({})
// @ts-ignore
Simple.args = {
  children: 'A alert that informs you of stuff',
}
