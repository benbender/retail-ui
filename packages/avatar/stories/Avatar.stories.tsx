import * as React from 'react'

import { Avatar, AvatarProps } from '../src'

const meta = {
  title: 'Retail-UI/Avatar',
  component: Avatar,
}

export default meta

const Template = (args: AvatarProps) => (
  <Avatar {...args} data-testid={meta.title} />
)

export const Simple = Template.bind({})
// @ts-ignore
Simple.args = {
  src: `https://pbs.twimg.com/profile_images/1285102166709919744/XvlWrJny_400x400.jpg`,
}
