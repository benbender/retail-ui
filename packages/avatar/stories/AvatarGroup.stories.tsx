import * as React from 'react'

import { Avatar, AvatarGroup, AvatarProps } from '../src'

const meta = {
  title: 'Retail-UI/AvatarGroup',
  component: AvatarGroup,
}

export default meta

const src = `https://pbs.twimg.com/profile_images/1285102166709919744/XvlWrJny_400x400.jpg`

const Group = (args: AvatarProps) => (
  <AvatarGroup data-testid={meta.title}>
    <Avatar src={src} {...args} />
    <Avatar title="PJ" size="base" {...args} />
    {/* <Avatar icon={UserSolid} {...args} /> */}
  </AvatarGroup>
)

export const Simple = Group.bind({})
// @ts-ignore
Simple.args = {}
