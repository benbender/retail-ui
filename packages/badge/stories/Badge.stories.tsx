import * as React from 'react'

import { Badge, BadgeProps } from '../src'

const meta = {
  title: 'Retail-UI/Badge',
  component: Badge,
}

export default meta

const Template = (args: BadgeProps) => (
  <Badge {...args} data-testid={meta.title} />
)

export const Simple = Template.bind({})
// @ts-ignore
Simple.args = {
  children: 'Badge',
}
