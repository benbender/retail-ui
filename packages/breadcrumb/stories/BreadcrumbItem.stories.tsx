import * as React from 'react'

import { Breadcrumb, BreadcrumbItem, BreadcrumbProps } from '../src'

const meta = {
  title: 'Retail-UI/BreadcrumbItem',
  component: BreadcrumbItem,
}

export default meta

const Template = (args: BreadcrumbProps) => (
  <Breadcrumb data-testid={meta.title}>
    <BreadcrumbItem {...args}>
      <a href="#">Dashboard</a>
    </BreadcrumbItem>
    <BreadcrumbItem {...args}>
      <a href="#">Project</a>
    </BreadcrumbItem>
    <BreadcrumbItem isActive {...args}>
      Export
    </BreadcrumbItem>
  </Breadcrumb>
)

export const Simple = Template.bind({})
// @ts-ignore
Simple.args = {
  children: 'Breadcrumb',
}
