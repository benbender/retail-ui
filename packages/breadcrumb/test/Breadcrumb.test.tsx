import { render } from '@retail-ui/test-utils'
import * as React from 'react'

import * as Breadcrumb from '../stories/BreadcrumbItem.stories'

describe('Breadcrumb', () => {
  it('renders correctly', () => {
    const { asFragment } = render(<Breadcrumb.Simple />)
    expect(asFragment()).toMatchSnapshot()
  })
})
