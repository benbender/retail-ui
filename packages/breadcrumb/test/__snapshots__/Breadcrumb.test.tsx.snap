// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumb renders correctly 1`] = `
<DocumentFragment>
  <ul
    class="flex flex-wrap py-1 px-3 list-none rounded relative whitespace-no-wrap mt-0 mb-0"
    data-testid="Retail-UI/BreadcrumbItem"
  >
    <li
      class="text-gray-600 flex items-center mt-0 mb-0  text-base"
    >
      <span>
        <a
          href="#"
        >
          Dashboard
        </a>
      </span>
    </li>
    <span
      class="inline-block px-3 text-gray-600 last:hidden"
    >
      /
    </span>
    <li
      class="text-gray-600 flex items-center mt-0 mb-0  text-base"
    >
      <span>
        <a
          href="#"
        >
          Project
        </a>
      </span>
    </li>
    <span
      class="inline-block px-3 text-gray-600 last:hidden"
    >
      /
    </span>
    <li
      class="text-gray-600 flex items-center mt-0 mb-0  text-gray-900 no-underline text-base"
    >
      <span>
        Export
      </span>
    </li>
    <span
      class="inline-block px-3 text-gray-600 last:hidden"
    >
      /
    </span>
  </ul>
</DocumentFragment>
`;
